### ISSUE TITLE
Infinity loop in `Style/YodaCondition` autofix

### ISSUE THREAD

**Issue by ShockwaveNN** [bbatsov/rubocop#4412]
## Steps to reproduce the problem
1. Create `temp.rb` file
```
p 'a' if :a == :b
```
2. Run `rubocop -a temp.rb`. This cause exception
`Infinite loop detected in /path/test.rb`

## RuboCop version
```
$ rubocop -V
0.49.0 (using Parser 2.4.0.0, running on ruby 2.1.10 x86_64-linux)
```

**Comment by bbatsov** [comment]
Seems we didn't account for situations where both sides of an expression are some literals. :-)

**Comment by martinmosegaard** [comment]
I can also reproduce it with this code:

```
def test_yoda
  list = [{ number: 1 }]
  list.count { |elm| elm.number == 1 } > 0
end
```

- It reports a violation, although I would not like to have the comparison start with `0 < ...`
- It crashes with `check_for_infinite_loop` in autocorrect mode

**Comment by smakagon** [comment]
@bbatsov my bad, missed that case. Thanks guys for fixing that really quick.

**Comment by eagletmt** [comment]
I've found another false-positive case `self == var` and this case is also fixed by the commit 5700720cb92521e3a673e91ba517cc7884822826 .
 
@bbatsov Could you release a new version? This bug looks critical to me.

**Comment by bbatsov** [comment]
I'll likely do this after the weekend. I don't view the bug as critical as it affects a single cop which can easily be disabled.

**Comment by dankohn** [comment]
The following code produces YodaCondition errors in both directions:
```ruby
module StringRefinements
  refine String do
    def met?
      self == 'Met'
    end
  end
end
```
Here are the errors:
```
rubocop app/models/concerns/string_refinements.rb 
Inspecting 1 file
C
1  Style/YodaCondition
--
1  Total



Offenses:

app/models/concerns/string_refinements.rb:6:7: C: Style/YodaCondition: Reverse the order of the operands 'Met' == self. (https://en.wikipedia.org/wiki/Yoda_conditions)
      'Met' == self
      ^^^^^^^^^^^^^

1 file inspected, 1 offense detected

$ rubocop app/models/concerns/string_refinements.rb 
Inspecting 1 file
C
1  Style/YodaCondition
--
1  Total



Offenses:

app/models/concerns/string_refinements.rb:6:7: C: Style/YodaCondition: Reverse the order of the operands self == 'Met'. (https://en.wikipedia.org/wiki/Yoda_conditions)
      self == 'Met'
      ^^^^^^^^^^^^^

1 file inspected, 1 offense detected
```
